File: java/client/src/main/java/glide/api/models/commands/FailoverOptions.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.commands;

import java.util.ArrayList;
import java.util.List;
import lombok.NonNull;

/**
 * Optional arguments for {@link glide.api.GlideClient#failover(FailoverOptions)} and {@link
 * glide.api.GlideClusterClient#failover(FailoverOptions, glide.api.models.configuration.Route)}.
 *
 * @see <a href="https://valkey.io/commands/failover/">valkey.io</a> for details.
 */
public class FailoverOptions {

    private String host;
    private int port;
    private boolean force;
    private boolean abort;
    private Long timeout;

    private FailoverOptions() {}

    /**
     * Creates options to abort an ongoing failover.
     *
     * @return A new {@link FailoverOptions} with ABORT set.
     */
    public static FailoverOptions abort() {
        FailoverOptions opts = new FailoverOptions();
        opts.abort = true;
        return opts;
    }

    /**
     * Creates options with only a timeout.
     *
     * @param timeoutMs The maximum time in milliseconds to wait before aborting the failover.
     * @return A new {@link FailoverOptions} with TIMEOUT set.
     */
    public static FailoverOptions timeout(long timeoutMs) {
        FailoverOptions opts = new FailoverOptions();
        opts.timeout = timeoutMs;
        return opts;
    }

    /**
     * Creates options to failover to a specific replica.
     *
     * @param host The host of the target replica.
     * @param port The port of the target replica.
     * @return A new {@link FailoverOptions} targeting the specified replica.
     */
    public static FailoverOptions to(@NonNull String host, int port) {
        FailoverOptions opts = new FailoverOptions();
        opts.host = host;
        opts.port = port;
        return opts;
    }

    /**
     * Creates options to failover to a specific replica with a timeout.
     *
     * @param host The host of the target replica.
     * @param port The port of the target replica.
     * @param timeoutMs The maximum time in milliseconds to wait before aborting the failover.
     * @return A new {@link FailoverOptions} targeting the specified replica with a timeout.
     */
    public static FailoverOptions to(@NonNull String host, int port, long timeoutMs) {
        FailoverOptions opts = new FailoverOptions();
        opts.host = host;
        opts.port = port;
        opts.timeout = timeoutMs;
        return opts;
    }

    /**
     * Creates options to force a failover to a specific replica after the timeout elapses.
     *
     * @param host The host of the target replica.
     * @param port The port of the target replica.
     * @param timeoutMs The maximum time in milliseconds to wait before forcing the failover.
     * @return A new {@link FailoverOptions} with TO, TIMEOUT, and FORCE set.
     */
    public static FailoverOptions forced(@NonNull String host, int port, long timeoutMs) {
        FailoverOptions opts = new FailoverOptions();
        opts.host = host;
        opts.port = port;
        opts.force = true;
        opts.timeout = timeoutMs;
        return opts;
    }

    /**
     * Converts the options to command arguments.
     *
     * @return The command arguments as a string array.
     */
    public String[] toArgs() {
        List<String> args = new ArrayList<>();
        if (abort) {
            args.add("ABORT");
        } else {
            if (host != null) {
                args.add("TO");
                args.add(host);
                args.add(Integer.toString(port));
                if (force) {
                    args.add("FORCE");
                }
            }
            if (timeout != null) {
                args.add("TIMEOUT");
                args.add(Long.toString(timeout));
            }
        }
        return args.toArray(new String[0]);
    }
}

File: go/options/failover_options.go
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

package options

import "strconv"

// FailoverOptions represents options for the FAILOVER command.
type FailoverOptions struct {
	Host      string
	Port      int
	Force     bool
	Abort     bool
	TimeoutMs int64
}

// NewFailoverOptionsWithAbort creates options to abort an ongoing failover.
func NewFailoverOptionsWithAbort() *FailoverOptions {
	return &FailoverOptions{Abort: true}
}

// NewFailoverOptionsWithTimeout creates options with only a timeout.
func NewFailoverOptionsWithTimeout(timeoutMs int64) *FailoverOptions {
	return &FailoverOptions{TimeoutMs: timeoutMs}
}

// NewFailoverOptionsWithTo creates options to failover to a specific replica.
func NewFailoverOptionsWithTo(host string, port int) *FailoverOptions {
	return &FailoverOptions{Host: host, Port: port}
}

// NewFailoverOptionsWithToAndTimeout creates options to failover to a specific replica with a timeout.
func NewFailoverOptionsWithToAndTimeout(host string, port int, timeoutMs int64) *FailoverOptions {
	return &FailoverOptions{Host: host, Port: port, TimeoutMs: timeoutMs}
}

// NewFailoverOptionsForced creates options to force failover to a specific replica after timeout.
func NewFailoverOptionsForced(host string, port int, timeoutMs int64) *FailoverOptions {
	return &FailoverOptions{Host: host, Port: port, Force: true, TimeoutMs: timeoutMs}
}

// ToArgs converts the options to command arguments.
func (o *FailoverOptions) ToArgs() []string {
	args := []string{}
	if o.Abort {
		args = append(args, "ABORT")
		return args
	}
	if o.Host != "" && o.Port > 0 {
		args = append(args, "TO", o.Host, strconv.Itoa(o.Port))
		if o.Force {
			args = append(args, "FORCE")
		}
	}
	if o.TimeoutMs > 0 {
		args = append(args, "TIMEOUT", strconv.FormatInt(o.TimeoutMs, 10))
	}
	return args
}
